Userland/Libraries/LibWeb/XHR/FormData.cpp

```diff
@@ -0,0 +1,170 @@
+    // 2. Return the value of the first entry whose name is name from this’s entry list.
+    return m_entry_list.get(name)->at(0);
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-getall
+Vector<Variant<JS::NonnullGCPtr<FileAPI::File>, DeprecatedString>> FormData::get_all(DeprecatedString const& name)
+{
+    // 1. If there is no entry whose name is name in this’s entry list, then return the empty list.
+    if (!m_entry_list.contains(name))
+        return {};
+    // 2. Return the values of all entries whose name is name, in order, from this’s entry list.
+    return *m_entry_list.get(name);
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-has
+bool FormData::has(DeprecatedString const& name)
+{
+    // The has(name) method steps are to return true if there is an entry whose name is name in this’s entry list; otherwise false.
+    return m_entry_list.contains(name);
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-set
+WebIDL::ExceptionOr<void> FormData::set(DeprecatedString const& name, DeprecatedString const& value)
+{
+    auto& vm = realm().vm();
+    return inner_set(TRY_OR_THROW_OOM(vm, String::from_deprecated_string(name)), TRY_OR_THROW_OOM(vm, String::from_deprecated_string(value)));
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-set-blob
+WebIDL::ExceptionOr<void> FormData::set(DeprecatedString const& name, JS::NonnullGCPtr<FileAPI::Blob> const& blob_value, Optional<DeprecatedString> const& filename)
+{
+    auto& vm = realm().vm();
+    auto inner_filename = filename.has_value() ? TRY_OR_THROW_OOM(vm, String::from_deprecated_string(filename.value())) : Optional<String> {};
+    return inner_set(TRY_OR_THROW_OOM(vm, String::from_deprecated_string(name)), blob_value, inner_filename);
+}
+
+// https://xhr.spec.whatwg.org/#dom-formdata-set
+// https://xhr.spec.whatwg.org/#dom-formdata-set-blob
+WebIDL::ExceptionOr<void> FormData::inner_set(String const& name, Variant<JS::NonnullGCPtr<FileAPI::Blob>, String> const& value, Optional<String> const& filename)
+{
+    auto& realm = this->realm();
+
+    // 1. Let value be value if given; otherwise blobValue.
+    // 2. Let entry be the result of creating an entry with name, value, and filename if given.
+    auto entry = TRY(HTML::create_entry(realm, name, value, filename));
+
+    // FIXME: Remove this when our binding generator supports "new string".
+    auto form_data_entry_value = entry.value.has<String>()
+        ? FormDataEntryValue { entry.value.get<String>().to_deprecated_string() }
+        : FormDataEntryValue { entry.value.get<JS::NonnullGCPtr<FileAPI::File>>() };
+
+    // 3. If there are entries in this’s entry list whose name is name, then replace the first such entry with entry and remove the others.
+    if (auto entries = m_entry_list.get(entry.name.to_deprecated_string()); entries.has_value() && !entries->is_empty()) {
+        entries->remove(0, entries->size());
+        TRY_OR_THROW_OOM(realm.vm(), entries->try_append(form_data_entry_value));
```

Comment:
please `auto& vm =` once at the beginning to reduce visual noise & cognitive overhead - where the VM is from never matters :)